compass/landice/tests/ensemble_generator/branch_ensemble/branch_run.py

```diff
@@ -0,0 +1,184 @@
+
+        # create restart_timestamp
+        with open(os.path.join(self.work_dir, 'restart_timestamp'), 'w') as f:
+            f.write('2015-01-01_00:00:00')
+
+        # yaml file
+        shutil.copy(os.path.join(spinup_dir, 'albany_input.yaml'),
+                    self.work_dir)
+
+        # set up namelist
+        # start with the namelist from the spinup
+        # Note: this differs from most compass tests, which would start with
+        # the default namelist from the mpas build dir
+        namelist = compass.namelist.ingest(os.path.join(spinup_dir,
+                                                        'namelist.landice'))
+        # use the namelist in this module to update the spinup namelist
+        options = compass.namelist.parse_replacements(
+            'compass.landice.tests.ensemble_generator.branch_ensemble',
+            'namelist.landice')
+        namelist = compass.namelist.replace(namelist, options)
+        compass.namelist.write(namelist, os.path.join(self.work_dir,
+                                                      'namelist.landice'))
+
+        # set up streams
+        stream_replacements = {}
+        TF_file_path = section.get('TF_file_path')
+        stream_replacements['TF_file_path'] = TF_file_path
+        SMB_file_path = section.get('SMB_file_path')
+        stream_replacements['SMB_file_path'] = SMB_file_path
+        strm_src = 'compass.landice.tests.ensemble_generator.branch_ensemble'
+        self.add_streams_file(strm_src,
+                              'streams.landice',
+                              out_name='streams.landice',
+                              template_replacements=stream_replacements)
+
+        # copy run_info file
+        shutil.copy(os.path.join(spinup_dir, 'run_info.cfg'), self.work_dir)
+
+        # copy graph files
+        shutil.copy(os.path.join(spinup_dir, 'graph.info'), self.work_dir)
+
+        # save number of tasks to use
+        # eventually compass could determine this, but for now we want
+        # explicit control
+        self.ntasks = self.config.getint('ensemble', 'ntasks')
+        self.min_tasks = self.ntasks
+
+        self.add_model_as_input()
+
+        # set job name to run number so it will get set in batch script
+        # Note: currently, for this to work right, one has to delete/comment
+        # the call to write_job_script at line 316-7 in compass/setup.py
+        self.config.set('job', 'job_name', f'uq_{self.name}')
+        machine = self.config.get('deploy', 'machine')
+        pre_run_cmd = ('LOGDIR=previous_logs_`date +"%Y-%m-%d_%H-%M-%S"`;'
+                       'mkdir $LOGDIR; cp log* $LOGDIR; date')
+        post_run_cmd = "date"
+        write_job_script(self.config, machine,
+                         target_cores=self.ntasks, min_cores=self.min_tasks,
+                         work_dir=self.work_dir,
+                         pre_run_commands=pre_run_cmd,
+                         post_run_commands=post_run_cmd)
```

Comment:
@xylar , when setting up ensembles, we ideally would be able to customize the job script to have the name of the run, as well as these pre and post run commands.  However, because `compass setup` writes each step's job script (https://github.com/MPAS-Dev/compass/blob/main/compass/setup.py#L315) , **after** the step has been set up (https://github.com/MPAS-Dev/compass/blob/main/compass/setup.py#L295), whatever job script gets written here in the step's `setup` method gets overwritten.  This means that the pre and post commands here get lost, and all ensemble members have the job name of the last processed ensemble member (e.g. when submitting a 100 run ensemble, in looking at squeue, all jobs have the name `run100`)
The ensemble_generator does not follow normal compass protocols, so maybe there is not a good workaround to this situation, but I was wondering if you had any ideas.

The ideas I have considered are:
1. somehow being able to disable the writing of job script for a step (not sure how that would be controlled), so that the clobbering wouldn't happen
2. being able to have a step's `setup` pass back arguments that can optionally be used by `write_job_script`
3. have a 'post-setup' method in a step that `compass setup` calls at the very end of `setup_case` where the job script could be overwritten

All of these solutions seem pretty awkward and invasive to solve a very specific situation.  Can you think of any other ideas?  It's not crucial that the job scripts get customized like this, but it does help a lot with managing large ensembles.

Reply:
I think the best solution is to create test-case-specific job script generator that uses a special job script name and then have the ensemble_manager use that.  I will save that for a different PR.